cpp/ql/src/Likely Bugs/Format/NonConstantFormat.ql

```diff
@@ -37,6 +37,40 @@ class UncalledFunction extends Function {
   }
 }
 
+/**
+ * Holds if `t` cannot refer to a string. That is, it's a built-in
+ * or arithmetic type that is not a "`char` like" type.
+ */
+predicate cannotContainString(Type t) {
+  exists(Type unspecified |
+    unspecified = t.getUnspecifiedType() and
+    not unspecified instanceof UnknownType and
+    not unspecified instanceof CharType and
+    not unspecified instanceof WideCharType and
+    not unspecified instanceof Char8Type and
+    not unspecified instanceof Char16Type and
+    not unspecified instanceof Char32Type and
+    // C often defines `wchar_t` as `unsigned short`
+    unspecified = any(ShortType short | not short.isUnsigned())
```

Comment:
> Though if all UserTypes are considered part of "canContainString" then I suppose "cannotContainString" is likely to be much smaller and thus more performant?

`UserType`s should indeed not be barriers since they can contain fields that contain strings.

> I think the logic of this line is wrong, as it is we're forcing unspecified to be a signed short.
Oh, I think you're right. Let me fix that and double-check performance on ImageMagick (since they have a lot of `wchar_t`s that are really `unsigned short`s).

Reply:
It may actually be the case that we can remove the "source root check" I introduced in 51f57407071cc09cbbc6707722a15bfc5f9828df. So please don't spend too much time looking at results yet before I know whether that slightly-more-risky change is actually needed

Reply:
Fixed in 61597f5ac79b2a38285a2299fc0d5439d6257ae0. I missed the fact that you suggested the fix in your comment, so I pushed an equivalent fix. Sorry about that!

As a bonus, because you spotted this problem, we no longer need the controversial "source root" change 🎉